test/integration/api_v1_test.rb

```diff
@@ -586,6 +586,27 @@ class ApiV1Test < ActionDispatch::IntegrationTest
     assert_response :unauthorized
   end
 
+  test 'add a key to a project' do
```

Comment:
Actually, the error is a good thing!!! Whenever you have a test that fails, you should modify your code so that it passes your test. This is called test-driven development. It's super popular these days (at startups especially), so this is a good learning experience for the both of us. So let's have a look:
```
@project = Project.find(params[:contrib_key][:project_id])

if @cur_user.id == @project.user_id
  session[:name] = params[:key_name]
  session[:key] = params[:key]
  session[:project_id] = @project.id
  key = ContribKey.new(contrib_key_params)
  key.save
  end
```
First things first: you must check if the things in params are nil!!!
This means that params[:contrib_key], params[:contrib_key][:project_id], params[:key_name], and params[:key] must all be checked non-nil.

Since we are special casing success, I would skip the if block if any of these values are nil. This way the respond to |format| block can remain the way it is.

Otherwise this patch is quite elegant.